interfaces/prompting/patterns/variant.go

```diff
@@ -0,0 +1,353 @@
+			} else {
+				components = append(components, component{compType: compAnySingle})
+			}
+		case '⁑':
+			consumeText()
+			if prevComponentsAre([]componentType{compSeparatorDoublestar, compSeparator}) {
+				// Reduce /**/** to simply /** by removing most recent separator
+				components = components[:len(components)-1]
+			} else if prevComponentsAre([]componentType{compSeparator}) {
+				// Replace previous separator with separatorDoublestar
+				components[len(components)-1] = component{compType: compSeparatorDoublestar}
+			} else {
+				// Reduce to * since previous component is not a separator
+				addGlobstar()
+			}
+		case '*':
+			reducePrevDoublestar()
+			consumeText()
+			addGlobstar()
+		case '\\':
+			r2, _, err := rr.ReadRune()
+			if err != nil {
+				// Should be impossible, we just rendered this variant
+				return nil, errors.New(`internal error: trailing unescaped '\' character`)
+			}
+			switch r2 {
+			case '*', '?', '[', ']', '{', '}', '\\':
+				runes = append(runes, r, r2)
+			default:
+				// do not add r to runes if it's unnecessary
+				runes = append(runes, r2)
+			}
+		case '[', ']', '{', '}':
+			// Should be impossible, we just rendered this variant
+			return nil, fmt.Errorf(`internal error: unexpected unescaped '%v' character`, r)
+		default:
+			runes = append(runes, r)
+		}
+	}
+
+	consumeText()
+
+	if prevComponentsAre([]componentType{compSeparatorDoublestar, compSeparator, compGlobstar}) {
+		// If components end with /**/*, strip trailing /*
+		components = components[:len(components)-2]
+	}
+
+	// Add terminal marker or convert existing doublestar to terminal doublestar
+	if prevComponentsAre([]componentType{compSeparatorDoublestar, compSeparator}) {
+		components = components[:len(components)-2]
+		components = append(components, component{compType: compSeparatorDoublestarSeparatorTerminal})
+	} else if prevComponentsAre([]componentType{compSeparatorDoublestar}) {
+		components[len(components)-1] = component{compType: compSeparatorDoublestarTerminal}
+	} else {
+		components = append(components, component{compType: compTerminal})
+	}
+
+	var variantBuf strings.Builder
+	var regexBuf strings.Builder
+	regexBuf.WriteRune('^')
+	for _, c := range components {
+		variantBuf.WriteString(c.String())
+		regexBuf.WriteString(c.componentRegex())
+	}
+	regexBuf.WriteRune('$')
+	regex := regexp.MustCompile(regexBuf.String())
+	regex.Longest()
+
+	v := PatternVariant{
+		variant:    variantBuf.String(),
+		components: components,
+		regex:      regex,
+	}
+
+	return &v, nil
+}
+
+// Need to replace unescaped "**", but must be careful about an escaped '\\'
+// before the first '*', since that doesn't escape the first '*'.
+var doublestarReplacer = regexp.MustCompile(`([^\\](\\\\)*)\*\*`)
+
+func prepareVariantForParsing(variant string) string {
+	prepared := doublestarReplacer.ReplaceAllStringFunc(variant, func(s string) string {
+		// Discard trailing "**", add "⁑" instead
+		return s[:len(s)-2] + "⁑"
```

Comment:
Nice :)

Is there anything that prevents us from seeing `⁑` in input that would upset some logic below?

Reply:
There's not, good catch! I'll add a validation to escape any literal `⁑` before doing this replacement.

Reply:
Furthermore, there are bugs here...

Specifically, correctly handling patterns like this: `/foo/**\\**\\**\\**\\**` or `/foo/⁑\\⁑\\⁑\\⁑\\⁑`. If there are multiple special character sequences in a row, the special character at the end of the one match can't be used as the leading `[^\\]` in the next match, so we skip a sequence.

Might need to resort to returning the start/end indices of the submatch, then checking if the preceding rune is the special character, without including that rune in the match itself.

Reply:
Fix in latest commit :sunglasses: but the parsing code is increasingly complex... I'll see what I can do.